src/metabase/driver/sql_jdbc/actions.clj

```diff
@@ -1,235 +1,277 @@
+    (log/tracef ":row/create HoneySQL:\n\n%s" (u/pprint-to-str create-hsql))
+    (log/tracef ":row/create SQL + args:\n\n%s" (u/pprint-to-str sql-args))
+    (with-jdbc-transaction [conn database-id]
+      (try
+        (let [result (jdbc/execute! {:connection conn} sql-args {:return-keys true, :identifiers identity, :transaction? false})
+              _      (log/tracef ":row/create INSERT returned\n\n%s" (u/pprint-to-str result))
+              row    (if (= driver :postgres)
+                       result
+                       (select-created-row driver raw-hsql conn result))]
+          (log/tracef ":row/create returned row %s" (pr-str row))
+          {:created-row row})
+        (catch Exception e
+          (catch-throw e 400 {:query      query
+                              :driver     driver
+                              :raw-hsql   raw-hsql
+                              :create-sql create-hsql
+                              :sql-args   sql-args}))))))
+
+(defmulti do-nested-transaction
+  "Execute `thunk` inside a nested transaction inside `connection`, which is currently in a transaction. If `thunk`
+  throws an Exception, the nested transaction should be rolled back, but the parent transaction should be able to
+  proceed."
+  {:arglists '([driver ^java.sql.Connection connection thunk]), :added "0.44.0"}
+  driver/dispatch-on-initialized-driver
+  :hierarchy #'driver/hierarchy)
+
+(defmethod actions/perform-action!* [:sql-jdbc :bulk/create]
+  [driver _action {database-id :id, :as database} {:keys [table-id rows]}]
+  (log/tracef "Inserting %d rows" (count rows))
+  (with-jdbc-transaction [conn database-id]
+    (transduce
+     (map-indexed (fn [i row] [i row]))
+     (fn
+       ([]
+        [[] []])
+       ([[errors created-rows]]
+        ;; if there were any errors throw an Exception with all the error messages.
+        (when (seq errors)
+          (.rollback conn)
+          (throw (ex-info (tru "Error(s) inserting rows.")
+                          {:status-code 400, :errors errors})))
+        ;; if there we no errors then return the created rows.
+        {:created-rows created-rows})
+       ([[errors created-rows] [row-index row]]
+        (try
+          (let [result (do-nested-transaction
+                        driver
+                        conn
+                        (fn []
+                          (actions/perform-action!*
+                           driver
+                           :row/create
+                           database
+                           {:database   database-id
+                            :type       :query
+                            :query      {:source-table table-id}
+                            :create-row row})))]
+            [errors (conj created-rows (:created-row result))])
+          (catch Throwable e
+            [(conj errors {:index row-index, :error (ex-message e)})
+             created-rows]))))
+     rows)))
```

Comment:
It's not too clear why nested transactions are necessary.

I believe it's because: if row #1 fails, normally during a transaction then subsequent rows would fail with a "transaction rolled back" error even if those rows would normally succeed. This allows collecting "real" errors and then rolling back the whole thing at the end.

Maybe a comment explaining that would help.